interactions/api/http.py

```diff
@@ -0,0 +1,2285 @@
+            self.session = ClientSession()
+
+    async def request(self, route: Route, **kwargs) -> Optional[Any]:
+        r"""
+        Sends a request to the Discord API.
+
+        :param route: The HTTP route to request.
+        :type route: Route
+        :param \**kwargs: Optional keyword-only arguments to pass as information in the request.
+        :type \**kwargs: dict
+        :return: The contents of the request if any.
+        :rtype: Optional[Any]
+        """
+        self.check_session()
+
+        bucket: Optional[str] = route.bucket
+
+        for _ in range(3):
+            ratelimit: Lock = self.ratelimits.get(bucket)
+
+            if not self.lock.is_set():
+                log.warning("Global lock is still locked, waiting for it to clear...")
+                await self.lock.wait()
+
+            if ratelimit is None:
+                self.ratelimits[bucket] = Lock()
+                continue
+
+            await ratelimit.acquire()
+
+            with Padlock(ratelimit) as lock:  # noqa: F841
+                kwargs["headers"] = {**self.headers, **kwargs.get("headers", {})}
+
+                try:
+                    reason = kwargs.get("reason")
+                except:  # noqa
+                    pass
+                else:
+                    if reason:
+                        kwargs["headers"]["X-Audit-Log-Reason"] = quote(reason, safe="/ ")
+
+                kwargs["headers"]["Content-Type"] = "application/json"
+                async with self.session.request(
+                    route.method, route.__api__ + route.path, **kwargs
+                ) as response:
+                    if not (response.headers["X-Ratelimit-Remaining"] - self.limit):
+                        log.info(
+                            "The application has reached the last remaining request. Cooling down to avoid rate limiting."
+                        )
+                        self.lock.set()
+                        await sleep(1)
+                        self.lock.clear()
+                    else:
+                        self.limit = response.headers["X-Ratelimit-Remaining"]
+                        data = await response.json(content_type=None)
+                        log.debug(data)
+
+                    if response.status in (300, 401, 403, 404):
+                        raise HTTPException(response.status)
+                    elif response.status == 429:
+                        retry_after = data["retry_after"]
+
+                        if "X-Ratelimit-Global" in response.headers.keys():
+                            self.lock.set()
+                            log.warning("The HTTP request has encountered a global API ratelimit.")
+                            await sleep(retry_after)
+                            self.lock.clear()
+                        else:
+                            log.warning("A local ratelimit with the bucket has been encountered.")
+                            await sleep(retry_after)
```

Comment:
This should also lock the per route ratelimit

Reply:
This already should be doing that